packages/playwright-crawler/src/internals/adaptive-playwright-crawler.ts

```diff
@@ -0,0 +1,263 @@
+        const id = this.idOrDefault(idOrName);
+        this.keyValueStoreChanges[id] ??= {};
+        return this.keyValueStoreChanges[id][key]?.changedValue ?? null;
+    };
+
+    private setKeyValueStoreChangedValue = (idOrName: string | undefined, key: string, changedValue: unknown, options?: RecordOptions) => {
+        const id = this.idOrDefault(idOrName);
+        this.keyValueStoreChanges[id] ??= {};
+        this.keyValueStoreChanges[id][key] = { changedValue, options };
+    };
+
+    getKeyValueStore: RestrictedCrawlingContext['getKeyValueStore'] = async (idOrName) => {
+        const store = await KeyValueStore.open(idOrName, { config: this.config });
+
+        return {
+            id: this.idOrDefault(idOrName),
+            name: idOrName,
+            getValue: async (key) => this.getKeyValueStoreChangedValue(idOrName, key) ?? await store.getValue(key),
+            getAutoSavedValue: async <T extends Dictionary = Dictionary>(key: string, defaultValue: T = {} as T) => {
+                let value = this.getKeyValueStoreChangedValue(idOrName, key);
+                if (value === null) {
+                    value = await store.getValue(key) ?? defaultValue;
+                    this.setKeyValueStoreChangedValue(idOrName, key, value);
+                }
+
+                return value as T;
+            },
+            setValue: async (key, value, options) => {
+                this.setKeyValueStoreChangedValue(idOrName, key, value, options);
+            },
+        };
+    };
+}
+
+interface AdaptivePlaywrightCrawlerContext extends RestrictedCrawlingContext {
+    dom: CheerioPortadom<Cheerio<Document>> | PlaywrightLocatorPortadom;
+}
+
+interface AdaptivePlaywrightCrawlerOptions extends Omit<PlaywrightCrawlerOptions, 'requestHandler'> {
+    requestHandler: (crawlingContext: AdaptivePlaywrightCrawlerContext) => Awaitable<void>;
+    renderingTypeDetectionRatio: number;
+    resultChecker?: (result: RequestHandlerResult) => boolean;
+    resultComparator?: (resultA: RequestHandlerResult, resultB: RequestHandlerResult) => boolean;
+}
+
+export class AdaptivePlaywrightCrawler extends PlaywrightCrawler {
+    private adaptiveRequestHandler: AdaptivePlaywrightCrawlerOptions['requestHandler'];
+    private renderingTypePredictor: RenderingTypePredictor;
+    private resultChecker: NonNullable<AdaptivePlaywrightCrawlerOptions['resultChecker']>;
+    private resultComparator: NonNullable<AdaptivePlaywrightCrawlerOptions['resultComparator']>;
+
+    constructor({ requestHandler, renderingTypeDetectionRatio, resultChecker, resultComparator, ...options }: AdaptivePlaywrightCrawlerOptions) {
+        super(options);
+        this.adaptiveRequestHandler = requestHandler;
+        this.renderingTypePredictor = new RenderingTypePredictor({ detectionRatio: renderingTypeDetectionRatio });
+        this.resultChecker = resultChecker ?? (() => true);
+        this.resultComparator = resultComparator ?? (() => true); // TODO
```

Comment:
I'll make a default comparator after I figure out a better API for the `RequestHandlerResult`

Reply:
I imagine that if there is a `resultChecker`, the comparator will check both results. If there is not, it will check both datasets for deep equality.